zebrad/src/commands/migrate_from_zcashd.rs

```diff
@@ -0,0 +1,371 @@
+//!    * writes best finalized chain blocks to permanent storage,
+//!      in the new format
+//!    * only performs essential contextual verification of blocks,
+//!      to make sure that block data hasn't been corrupted by
+//!      receiving blocks in the new format
+//!      * TODO: We currently perform full validaton because we don't read from the
+//!        `zcashd` block index.
+//!    * fetches blocks from the best finalized chain from permanent storage,
+//!      in the new format
+
+use std::{
+    path::{Path, PathBuf},
+    sync::Arc,
+};
+
+use abscissa_core::{config, Command, FrameworkError};
+use color_eyre::eyre::{eyre, Report};
+use tokio::{
+    fs::File,
+    io::{AsyncReadExt, BufReader},
+    time::Instant,
+};
+use tower::{Service, ServiceExt};
+
+use zebra_chain::{
+    block::{Block, Height},
+    parameters::{Magic, Network},
+    serialization::ZcashDeserialize,
+};
+
+use crate::{
+    components::tokio::{RuntimeRun, TokioComponent},
+    config::ZebradConfig,
+    prelude::*,
+    BoxError,
+};
+
+/// How often we log info-level progress messages
+const PROGRESS_HEIGHT_INTERVAL: u32 = 5_000;
+
+/// Creates a new Zebra state from an existing `zcashd` datadir
+#[derive(Command, Debug, clap::Parser)]
+pub struct MigrateFromZcashdCmd {
+    /// `zcashd` datadir from which to migrate state.
+    #[clap(long)]
+    datadir: PathBuf,
+
+    /// Source height that the migration finishes at.
+    #[clap(long, short, help = "stop copying at this source height")]
+    max_source_height: Option<u32>,
+
+    /// Filter strings which override the config file and defaults
+    #[clap(help = "tracing filters which override the zebrad.toml config")]
+    filters: Vec<String>,
+}
+
+impl MigrateFromZcashdCmd {
+    /// Configure and launch the migrate command
+    async fn start(&self) -> Result<(), Report> {
+        let app_config = APPLICATION.config();
+
+        self.migrate(app_config.network.network.clone(), app_config.state.clone())
+            .await
+            .map_err(|e| eyre!(e))
+    }
+
+    /// Initialize the target state, then copy from the `zcashd` datadir to the target
+    /// state.
+    async fn migrate(
+        &self,
+        network: Network,
+        target_config: zebra_state::Config,
+    ) -> Result<(), BoxError> {
+        info!(?target_config, "initializing target state service");
+
+        let target_start_time = Instant::now();
+        // We're not verifying UTXOs here, so we don't need the maximum checkpoint height.
+        //
+        // TODO: call Options::PrepareForBulkLoad()
+        // See "What's the fastest way to load data into RocksDB?" in
+        // https://github.com/facebook/rocksdb/wiki/RocksDB-FAQ
+        let (
+            mut target_state,
+            _target_read_only_state_service,
+            _target_latest_chain_tip,
+            _target_chain_tip_change,
+        ) = zebra_state::spawn_init(target_config.clone(), &network, Height::MAX, 0).await?;
```

Comment:
I'm skeptical of there being any noticeable benefit from using `PrepareForBulkLoad()` until we switch to shardtrees, but it would be nice to have.

Note: The transaction verifier works with a dummy mempool setup channel receiver.

```suggestion
        let (
            target_state_service,
            mut target_read_only_state_service,
            _target_latest_chain_tip,
            _target_chain_tip_change,
        ) = zebra_state::spawn_init(target_state_config.clone(), &network, Height::MAX, 0).await?;
        
        let target_state = ServiceBuilder::new()
            .buffer(STATE_BUFFER_BOUND)
            .service(target_state_service);

        let (
            mut block_verifier_router,
            _tx_verifier,
            _consensus_task_handles,
            _max_checkpoint_height,
        ) = zebra_consensus::router::init(
            target_consensus_config,
            &network,
            target_state,
            oneshot::channel::<
                Buffer<BoxService<mempool::Request, mempool::Response, BoxError>, mempool::Request>,
            >()
            .1,
        )
        .await;

```